### PR TITLE
Fix :endclass :endenum :endinterface giving errors

### PR DESCRIPTION
**Pull request by kennypete**
This fixes `:endclass`, `:endenum`, and `:endinterface` from giving errors, which appears to have been unintentional.  [The errors currently, with screenshots, and discussion on the matter can be found in #20191]

The test file `src/testdir/test_vim9_class.vim` is updated to test these as well as adding basic tests for enum and interface.  They are also in 20191, which can easily be aligned after this PR is merged.

### CODE REVIEW

File: src/vim9class.c [hunk 1]
```diff
@@ -2218,9 +2218,14 @@ ex_class(exarg_T *eap)
...
+	// Handle a preceding colon (otherwise :endclass :endenum and
+	// :endinterface give errors)
+	if (*p == ':')
+	    p = skipwhite(p + 1);
+	char_u *cmd_start = p;
 	if (checkforcmd(&p, end_name, shortlen))
 	{
-	    if (STRNCMP(line, end_name, fullen) != 0)
+	    if (STRNCMP(cmd_start, end_name, fullen) != 0)
 		semsg(_(e_command_cannot_be_shortened_str), line);
```
Comment (dkearns) [inline src/vim9class.c]: - I think this should use `cmd_start` for the error message to exclude any leading context.
- These commands should allow multiple colons with arbitrary interspersed whitespace.   Something like:
https://github.com/vim/vim/blob/0fa3603d6d9cec774f3060a33c59778c9baf0e41/src/userfunc.c#L1120-L1122
- While they should, in principle, also emit "No range allowed" messages if one is specified some of the other block closing commands (:end{func,def} at least) don't emit these either.

**Comment by dkearns** [comment]
Thanks, this was little irksome. :)

File: src/vim9class.c (hunk 1 above)
Comment (kennypete) [reply src/vim9class.c]: Fixed, and tests have been added for them, which run clean. I also added tests with colon+whitespace variants, which revealed `:public` gives E1065, though that can wait for another day (and the test has a TODO and `: public` commented).

File: src/testdir/test_vim9_class.vim [hunk 2]
```diff
@@ -11869,4 +11897,176 @@ func Test_class_member_lambda()
   call v9.CheckSourceSuccess(lines)
 endfunc
 
+def Test_enum_basic()
```
Comment (dkearns) [inline src/testdir/test_vim9_class.vim]: Is there a reason not to put this in test_vim9_enum.vim?

File: src/testdir/test_vim9_class.vim [hunk 3]
```diff
@@ -11869,4 +11897,176 @@ func Test_class_member_lambda()
...
+    vim9script
+    enum NoAdditionalCmd
+      One, | var y = 10
+      Two
+    endenum
+  END
+  v9.CheckSourceFailure(lines, "E1418: Invalid enum value declaration: | var y = 10", 3)
+enddef
+
+def Test_interface_basic()
```
Comment (dkearns) [inline src/testdir/test_vim9_class.vim]: ...and this in test_vim9_interface.vim?

File: src/vim9class.c [hunk 4]
```diff
@@ -2230,10 +2230,15 @@ ex_class(exarg_T *eap)
 	    fullen = 12;
 	}
 
+	// Handle a preceding colon (otherwise :endclass :endenum and
+	// :endinterface give errors)
```
Comment (dkearns) [inline src/vim9class.c]: I think this comment mostly captures the motivation for the fix which is best left to the commit message.  I don't think it really needs a comment but get_function_body() used:

```suggestion
	// skip ':' and blanks
```

Something like the following (untested and mainly to point you at skip_range()) could be used to handle illegal ranges if you wanted to address this at the same time:

```diff
diff --git a/src/vim9class.c b/src/vim9class.c
index 97ef51116..43b758029 100644
--- a/src/vim9class.c
+++ b/src/vim9class.c
@@ -2234,9 +2234,19 @@ early_ret:
 	// :endinterface give errors)
 	for (; VIM_ISWHITE(*p) || *p == ':'; ++p)
 	    ;
+
+	char_u *maybe_range = p;
+	p = skip_range(p, TRUE, NULL);
 	char_u *cmd_start = p;
+
 	if (checkforcmd(&p, end_name, shortlen))
 	{
+	    if (cmd_start > maybe_range)
+	    {
+		emsg(_(e_no_range_allowed));
+		break;
+	    }
+
 	    if (STRNCMP(cmd_start, end_name, fullen) != 0)
 		semsg(_(e_command_cannot_be_shortened_str), cmd_start);
 	    else if (*p == '|' || !ends_excmd2(line, p))
```

**Review by dkearns — COMMENTED** [review]
Sorry for the slightly half-baked and spiralling reviews.  I'm a bit short on time at the moment and often just using the GitHub web interface.

File: src/testdir/test_vim9_class.vim (hunk 2 above)
Comment (kennypete) [reply src/testdir/test_vim9_class.vim]: None, just I'd not noticed it.  There may be some duplication, so it shouldn't be hard to move and deduplicate where appropriate.

File: src/testdir/test_vim9_class.vim (hunk 3 above)
Comment (kennypete) [reply src/testdir/test_vim9_class.vim]: Ditto test_vim9_enum.vim comment

File: src/vim9class.c (hunk 4 above)
Comment (kennypete) [reply src/vim9class.c]: I'll update the comment to `// skip ':' and blanks`; terse, fine.  On the range handling, I would prefer to park that and address it in a future PR.  Adding a range gives `E1318`, just not a range-specific error.  I want to conclude this PR now since it solves the core matter of `:end*` giving errors. Getting back to the (almost finished) big uplift of `vim9.txt` Section 2 is my aim.  That was interrupted by tracking down EX_WHOLE deficiencies (and later this spawned matter of `:end*` failing).  My time is better spent there too because Vim9 script is my thing, not C.

File: src/testdir/test_vim9_class.vim (hunk 2 above)
Comment (kennypete) [reply src/testdir/test_vim9_class.vim]: Deduped now: many tests were there already, though a few appeared to have been copied from test_vim9_class.vim and needed comment and/or name corrections. All tests clean.

**Comment by kennypete** [comment]
> Sorry for the slightly half-baked and spiralling reviews. I'm a bit short on time at the moment and often just using the GitHub web interface.

No worries.  Sorry for my slightly more than half-baked changes to that `test_vim9_class.vim` file, initially adding duplicate tests.  Feels like it's good to go now?  A future PR can pick up the couple of TODOs: 1) distinguishing the error given for adding a range, and 2) `:public` giving E1065.  I'm happy to do those, though after my `vim9.txt` Section 2 PR is merged.

File: src/testdir/test_vim9_enum.vim [hunk 5]
```diff
@@ -966,15 +993,15 @@ def Test_enum_whitespace()
   lines =<< trim END
     vim9script
     enum Car
-      Honda(), 
+      Honda(),
```
Comment (kennypete) [inline src/testdir/test_vim9_enum.vim:996]: Used a U+0009 tab here to test more than simply U+0020 space.

File: src/testdir/test_vim9_interface.vim [hunk 6]
```diff
@@ -281,13 +307,21 @@ def Test_class_implements_interface()
   END
   v9.CheckSourceFailure(lines, 'E1315: White space required after name: A"', 4)
 
-  # Trailing characters after a class name
+  # Trailing characters after an interface name
```
Comment (kennypete) [inline src/testdir/test_vim9_interface.vim:310]: This looks to have been a copy from the class test not updated when put into the **interface** test file; easily missed since it has the same error returned. Fixed now.

**Comment by kennypete** [comment]
Hi @chrisbra, this PR should be good to merge now. It's now a precursor to fixing EX_WHOLE deficiencies where the tests in 20191 (_which I've made _Draft_ because there are several things that need to be updated/fixed, including the additions and fix contained in this PR_) need to be included - noted at https://github.com/vim/vim/pull/20191#issuecomment-4529781503. Thanks

**Comment by chrisbra** [comment]
Thanks both!

**Comment by h-east** [comment]
This pull request is causing a CI error. Please address this issue.
https://github.com/vim/vim/actions/runs/26341481167/job/77544193058#step:8:43

**Comment by chrisbra** [comment]
Opps, sorry not sure how I missed that when merging yesterday. I'll fix it once I am back at my development machine
